File: latexdraw-research/GUImutants/mutant38/net.sf.latexdraw.mutant38/src/main/net/sf/latexdraw/actions/shape/MirrorShapes.scala
package net.sf.latexdraw.actions.shape

import net.sf.latexdraw.actions.Modifying
import net.sf.latexdraw.actions.DrawingAction
import net.sf.latexdraw.actions.ShapesAction
import org.malai.undo.Undoable
import org.malai.action.Action
import net.sf.latexdraw.actions.ShapeAction
import net.sf.latexdraw.glib.models.interfaces.IShape

/**
 * This action mirrors a shape.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2013 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2013-03-07<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
class MirrorShapes extends Action with ShapeAction[IShape] with Undoable with Modifying {
	/** If true, the mirroring is horizontal. */
	var _horizontally = true

	override protected def doActionBody() {
		val sh = _shape.get

		if(_horizontally)
			sh.mirrorHorizontal(sh.getGravityCentre)
		else
			sh.mirrorVertical(sh.getGravityCentre)
		sh.setModified(true)
	}


	override def canDo = _shape.isDefined


	override def undo() {
		doActionBody()
	}

	override def redo() {
		doActionBody()
	}

	override def getUndoName() = "mirror"

	override def isRegisterable()  = true

	/**
	 * If true, the mirroring is horizontal. Otherwise, vertical.
	 */
	def setHorizontally(horiz:Boolean) {
		_horizontally = horiz
	}
}

File: latexdraw-core/net.sf.latexdraw/src/main/java/net/sf/latexdraw/commands/shape/MirrorShapes.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2018 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.commands.shape;

import net.sf.latexdraw.commands.Modifying;
import net.sf.latexdraw.commands.ShapeCmdImpl;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import net.sf.latexdraw.util.LangTool;
import org.malai.undo.Undoable;

/**
 * This command mirrors a shape.
 * @author Arnaud Blouin
 */
public class MirrorShapes extends ShapeCmdImpl<IShape> implements Undoable, Modifying {
	/** If true, the mirroring is horizontal. */
	boolean horizontally;


	public MirrorShapes() {
		super();
		horizontally = true;
	}

	@Override
	protected void doCmdBody() {
		shape.ifPresent(sh -> {
			if(horizontally) {
				sh.mirrorHorizontal(sh.getGravityCentre().getX());
			}else {
				sh.mirrorVertical(sh.getGravityCentre().getY());
			}
			sh.setModified(true);
		});
	}


	@Override
	public boolean canDo() {
		return shape.isPresent();
	}

	@Override
	public void undo() {
		doCmdBody();
	}

	@Override
	public void redo() {
		doCmdBody();
	}

	@Override
	public String getUndoName() {
		return LangTool.INSTANCE.getBundle().getString("Actions.7");
	}

	@Override
	public RegistrationPolicy getRegistrationPolicy() {
		return RegistrationPolicy.LIMITED;
	}

	/**
	 * If true, the mirroring is horizontal. Otherwise, vertical.
	 */
	public void setHorizontally(final boolean horiz) {
		horizontally = horiz;
	}
}
